site/_posts/2023-01-12-what-happens-after-authentication-event.md

```diff
@@ -0,0 +1,136 @@
+* Advanced authentication functionality such as MFA, enterprise single sign-on and login rate limiting can be implemented in one place for all applications.
+* You can upgrade such authentication functionality without modifying downstream applications.
+* You can offer single sign-on across all your custom, commercial and open source applications.
+* Common login related workflows such as changing profile data or passwords can be centralized and managed by the auth server.
+
+Once you've decided to use the Authorization Code grant, you need to store the tokens that result from it. There are two main options. 
+
+## Store Tokens On The Client
+
+The first option is to send the access token and refresh token down to the client. 
+
+When using a browser, send these as `HTTPOnly`, secure cookies with a `SameSite` value of `Lax` or `Strict`.
+
+If you choose this option, the browser, whether a simple HTML page with some JavaScript or a complicated single page application (SPA), makes requests against APIs, and the token is along for the ride.
+
+As long as the APIs live on a common domain (or parent domain), the cookie will be sent. For example, the auth server can live at at `auth.example.com` and if you set the cookie domain to `.example.com`, APIs living at `api.example.com` and `todo.example.com`, or any other host under `.example.com`, will receive the token.
+
+{% plantuml source: _diagrams/blogs/after-authorization-code-grant/client-side-storage.plantuml, alt: "Storing the tokens as secure, HTTPOnly cookies." %}
+
+When using a native app, store these tokens in a secure location, such as the [iOS Keychain](https://developer.apple.com/documentation/security/keychain_services) or [Android internal data](https://developer.android.com/topic/security/best-practices#safe-data).
+
+### Token Validation
+
+Each API validates the token. They should check:
+
+* the signature
+* expiration time (the `exp` claim)
+* not valid before time (the `nbf` claim)
+* audience (the `aud` claim)
+* issuer (the `iss` claim)
+* any other specific claims
+
+All of these should be validated before handing back any data or allowing any functionality. In fact, they should be validated before any additional processing is done, because if any of these checks fail, the requester is unknown.
+
+The signature and standard claims checks can and should be done with a library. Checking other claims is business logic and should be handled by the API developer.
+
+This validation approach assumes the token has internal structure and is signed; a JWT meets these criteria and is used by FusionAuth and other auth servers.
+
+If you don't have a token that meets these criteria, introspect the token by presenting it to the auth server.
+
+{% plantuml source: _diagrams/blogs/after-authorization-code-grant/client-side-storage-introspection.plantuml, alt: "Storing the tokens as secure, HTTPOnly cookies and using introspection to validate them." %}
```

Comment:
It might be helpful to introduce the two main options at a high level in this diagram (or a new second diagram)

Reply:
@bhalsey do you mean the two main options to determine the validity of an access token?